### ISSUE TITLE
Better message for missing dependency?

### ISSUE THREAD

**Issue by thejohnfreeman** [ets-labs/python-dependency-injector#195]
If a `DependenciesContainer` is not overridden, or if its override is missing a provider, then attempts to call a missing provider will raise `Error('Dependency is not defined')`. Could we add the name of the missing provider?

**Comment by rmk135** [comment]
Hi @thejohnfreeman ,

It definitely makes sense to improve error message in this case, however, there is a little complexity to add this now, because providers are not bound to container. Generally, providers have no idea are they bound to any container or not, so this needs to be implemented carefully.

I will work on this, thanks for the question.

Thanks,
Roman

**Comment by wiryonolau** [comment]
It's difficult to trace error when the code have multiple layer of dependency

**Comment by shaunc** [comment]
This would definitely be useful -- could save a lot of debugging time. Perhaps, once we know there is an error, looking in the stack frame for context would be ok, even if not foolproof?

**Comment by rmk135** [comment]
Hey, I have released this improvement in version `4.21.0`. Here are a couple  of examples:

```python
import dataclasses

from dependency_injector import containers, providers


@dataclasses.dataclass
class UserService:
    database: object


class Container(containers.DeclarativeContainer):

    database = providers.Dependency()

    user_service = providers.Factory(
        UserService,
        database=database,
    )


if __name__ == '__main__':
    container = Container()
    container.user_service()

    # Message: Dependency "Container.database" is not defined
```

```python
from dependency_injector import containers, providers


class Database:
    ...


class UserService:
    def __init__(self, db: Database) -> None:
        self.db = db


class Gateways(containers.DeclarativeContainer):

    database_client = providers.Singleton(Database)


class Services(containers.DeclarativeContainer):

    gateways = providers.DependenciesContainer()

    user = providers.Factory(
        UserService,
        db=gateways.database_client,
    )


class Application(containers.DeclarativeContainer):

    gateways = providers.Container(Gateways)

    services = providers.Container(
        Services,
        # gateways=gateways,  # <---- missing dependency

    )


if __name__ == '__main__':
    application = Application()
    application.services().user()

    # Message: Dependency "Application.services.gateways.database_client" is not defined
```

```python
from dependency_injector import containers, providers


class Database:
    ...


class UserService:
    def __init__(self, db: Database) -> None:
        self.db = db


class Services(containers.DeclarativeContainer):

    gateways = providers.DependenciesContainer()

    user = providers.Factory(
        UserService,
        db=gateways.database_client,
    )


if __name__ == '__main__':
    services = Services()
    services.user()

    # Message: Dependency "Services.gateways.database_client" is not defined
```

Hope that it will help in development.

PS: I'm sorry it took almost 3 years to make it happen, but "later is better than never".

**Comment by shaunc** [comment]
Beautiful! :)
